### PR TITLE
D/sort day0

### PR DESCRIPTION
**Pull request by davidabrahams**

### CODE REVIEW

File: sorting/day00/README.md [hunk 1]
```diff
@@ -12,13 +12,31 @@ By doing this assignment, you should be able to:
   
 ## Assignment
 
+### Quicksort
```
Comment (davidabrahams) [inline sorting/day00/README.md]: As discussed, we won't have time to do QS in class. MIT OCW is nice :)

File: sorting/day00/README.md [hunk 2]
```diff
@@ -12,13 +12,31 @@ By doing this assignment, you should be able to:
...
+Feel free to watch at double speed!
+
+[Here](https://drive.google.com/open?id=0B_K4P69ad_l_RDY2U3VZUkpTY3M) are the corresponding powerpoint slides for the MIT lecture. [This](http://videolectures.net/mit6046jf05_leiserson_lec04/) website plays the video lecture side-by-side with the powerpoint lecture, with the downside of not being able to play faster than normal speed.
+
 ### Code
 
 - Draw the recursive trees (see [here](https://docs.google.com/presentation/d/1mNRb1tx3ibY2FjKkcXqJvjfDnT0vkxfYKEf38HrU26Y/edit?usp=sharing)) of the merge sort algorithm for the input array `[4,8,62,9,14,19]`. Show (a photo of) this to your NINJA at checkoff.
 - In `InsertionSort.java`, `MergeSort.java`, and `QuickSort.java`, complete the `sort` method.
-- For each `sort` function, detail the best, worst, and average case runtime complexity, as well as the space complexity, in the comments. Provide a short (one to two sentences total) explanation for the complexities.
+- For each `sort` function, detail the best, worst, and average case runtime complexity, as well as the space complexity, in the comments.
```
Comment (davidabrahams) [inline sorting/day00/README.md]: moved to NINJA explanation

File: sorting/day00/README.md [hunk 3]
```diff
@@ -12,13 +12,31 @@ By doing this assignment, you should be able to:
...
 
 ### Reading
 
 - **required**: When you're done coding, in *Think Algorithms* read section **23.2** and **23.5**.
-- **if you get stuck**: **23.3** and **23.4** contain implementations of insertion and merge sort. We encourage you to implement them without copying the book code.
\ No newline at end of file
+- **if you get stuck**: **23.3** and **23.4** contain implementations of insertion and merge sort. We encourage you to implement them without copying the book code.
```
Comment (davidabrahams) [inline sorting/day00/README.md]: idk i didnt change this line

File: sorting/day00/soln/MergeSort.java [hunk 4]
```diff
@@ -18,11 +17,13 @@
         int[] left = sort(tempLeft);
         int[] right = sort(tempRight);
 
-        merge(left, right, out);
-        return out;
+        return merge(left, right);
     }
 
-    private void merge(int[] left, int[] right, int[] out) {
+    public int[] merge(int[] left, int[] right) {
```
Comment (davidabrahams) [inline sorting/day00/soln/MergeSort.java]: matching header that we're giving the students

File: sorting/day00/src/MergeSort.java [hunk 5]
```diff
@@ -1,6 +1,8 @@
 
 public class MergeSort implements SortAlgorithm {
 
+    private static final int INSERTION_THRESHOLD = 10;
```
Comment (davidabrahams) [inline sorting/day00/src/MergeSort.java]: based on the WIKI page, 10 is commonly used? Figured we should give them this so they know to use it.

File: sorting/day00/soln/QuickSort.java [hunk 6]
```diff
@@ -24,7 +24,7 @@ private void quickSort(int[] a, int lo, int hi) {
      * than the pivot before the pivot, and all others after the pivot.
      * Return the index of the pivot in the array after partitioning.
      */
-    private int partition(int[] a, int lo, int hi) {
+    public int partition(int[] a, int lo, int hi) {
```
Comment (davidabrahams) [inline sorting/day00/soln/QuickSort.java]: public so we can test it

File: sorting/day00/src/QuickSort.java [hunk 7]
```diff
@@ -14,6 +16,16 @@
     }
 
     /**
+     * Partition the array around a pivot, then recursively sort the left and right
+     * portions of the array. A test for this method is provided in `SortTest.java`
+     * Optional: use Insertion Sort if the length of the array is <= INSERTION_THRESHOLD
+     */
+    public void quickSort(int[] a, int low, int high) {
```
Comment (davidabrahams) [inline sorting/day00/src/QuickSort.java]: i figured it might be a good idea to give them this intermediate method? I know it's almost the same as `sort`, but this way they can test that they're treating the boundaries correctly. (I wrote tests too)

File: sorting/day00/src/SortAlgorithm.java [hunk 8]
```diff
@@ -1,3 +1,11 @@
-public interface SortAlgorithm {
-    int[] sort(int[] array);
+public abstract class SortAlgorithm {
+
+    abstract int[] sort(int[] array);
+
+    static void swap(int[] a, int i, int j) {
```
Comment (davidabrahams) [inline sorting/day00/src/SortAlgorithm.java]: b/c both merge and insertion need this, I just moved it to the superclass

File: sorting/day00/tests/SortTest.java [hunk 9]
```diff
@@ -18,6 +21,8 @@
     @Before
     public void setup() {
         emptyCase = generateRandomArrayOfSize(0);
+        oneCase = generateRandomArrayOfSize(1);
+        twoCase = generateRandomArrayOfSize(2);
```
Comment (davidabrahams) [inline sorting/day00/tests/SortTest.java]: additional edge cases cant hurt

File: sorting/day00/tests/SortTest.java [hunk 10]
```diff
@@ -34,13 +39,16 @@ public void setup() {
...
+        assertArrayEquals(sorter.sort(oneCase), sort(oneCase));
+        assertArrayEquals(sorter.sort(twoCase), sort(twoCase));
         assertArrayEquals(sorter.sort(fastCase), sort(fastCase));
         assertArrayEquals(sorter.sort(slowCase), sort(slowCase));
     }
 
     private int[] sort(int[] array) {
-        Arrays.sort(array);
-        return array;
+        int[] copy = Arrays.copyOf(array, array.length);
```
Comment (davidabrahams) [inline sorting/day00/tests/SortTest.java]: Not sure this is necessary, but reading the tests I was a little nervous that they could pass because this `sort` function was sorting in place. Now, it leaves the original array alone (so we know their sort works)

File: sorting/day00/tests/SortTest.java [hunk 11]
```diff
@@ -63,6 +71,47 @@ public void testMerge() {
         assertArrayEquals(mergeSort.merge(a, b), a);
     }
 
+
+    @Test
+    public void testQuick() {
```
Comment (davidabrahams) [inline sorting/day00/tests/SortTest.java]: Tests for the intermediate quicksort function, which uses boundaries

File: sorting/day00/tests/SortTest.java [hunk 12]
```diff
@@ -63,6 +71,47 @@ public void testMerge() {
...
+        int[] a_sorted = sort(a);
+        quickSort.quickSort(a, 0, a.length-1);
+        assertArrayEquals(a, a_sorted);
+
+        a = new int[] {8, 3, 4, 6, 7, 2, 1, 2};
+        quickSort.quickSort(a, 1, a.length-2);
+        assertArrayEquals(a, new int[] {8, 1, 2, 3, 4, 6, 7, 2});
+
+        // Quicksort the middle of a an array
+        a = generateRandomArrayOfSize(984);
```
Comment (davidabrahams) [inline sorting/day00/tests/SortTest.java]: This is a bit of a shitshow....I wanted to basically call quicksort on a large subsection of an array. This was the best way I could think to test it. Feel free to modify.

File: sorting/day00/tests/SortTest.java [hunk 13]
```diff
@@ -63,6 +71,47 @@ public void testMerge() {
...
+        for (int i: a)
+            inputAL.add(i);
+        for (int i: b)
+            inputAL.add(i);
+        for (int i: c)
+            inputAL.add(i);
+        List<Integer> outputAL = new ArrayList<>();
+        for (int i: a)
+            outputAL.add(i);
+        for (int i: sort(b))
```
Comment (davidabrahams) [inline sorting/day00/tests/SortTest.java]: And here we make sure that the middle is sorted

File: sorting/day00/soln/MergeSort.java [hunk 14]
```diff
@@ -18,33 +17,27 @@
...
+
+        int[] out = new int[left.length + right.length];
+
         int i=0, j=0, k=0;
 
-        while (i < left.length && j < right.length) {
-            if (left[i] < right[j]) {
+        while (k < out.length) {
```
Comment (davidabrahams) [inline sorting/day00/soln/MergeSort.java]: This is a kinda nice way to not have to write two more while loops. You could argue it's less readable, so I left a comment. Not attached to this change

File: sorting/day00/soln/MergeSort.java [hunk 15]
```diff
@@ -18,33 +17,27 @@
...
+            // If j has reached the end of `right`, or i has not reached the end of `left` and left[i]<right[i],
+            // choose the element in left.
+            if (j==right.length || (i < left.length && left[i] < right[j])) {
```
Comment (davidabrahams) [inline sorting/day00/soln/MergeSort.java]: good call. refactored to helper method for readability
